utils/dependency_walker.go

```diff
@@ -0,0 +1,225 @@
+			}
+			if dc == len(node.dependencies) {
+				// All dependencies are there, place rule on the buffer
+				goodOrder = append(goodOrder, node.rule)  // Add to goodOrder buffer
+				nodes = append(nodes[:i], nodes[i+1:]...) // Delete from nodes
+				continue
+			}
+
+		}
+	}
+	rs.Rules = goodOrder
+	return rs
+}
+
+// GetDependenciesForRules will find all the dependencies (rules & modules)
+// for a slice of YARA rules
+func GetDependenciesForRules(ruleset ast.RuleSet, ruleNames ...string) (ast.RuleSet, error) {
+	// Make sure ruleNames and ruleset are not empty
+	if len(ruleset.Rules) == 0 {
+		return ast.RuleSet{}, fmt.Errorf("ruleset does not contain any rules")
+	}
+	// Make sure there are no duplicates in ruleset
+	var rulesetIdents []string
+	for _, rule := range ruleset.Rules {
+		if sliceContains(rule.Identifier, rulesetIdents) {
+			return ast.RuleSet{}, fmt.Errorf("ruleset contains a duplicate rule")
+		}
+		rulesetIdents = append(rulesetIdents, rule.Identifier)
+	}
+
+	// Ensure ruleNames are unique
+	ruleNamesM := make(map[string]struct{})
+	for _, ruleName := range ruleNames {
+		ruleNamesM[ruleName] = struct{}{}
+	}
+
+	var dependencies ast.RuleSet
+	for ruleName := range ruleNamesM {
+		var rule ast.Rule
+		for i := range ruleset.Rules {
+			if ruleset.Rules[i].Identifier == ruleName {
+				rule = *ruleset.Rules[i]
+				break
+			}
+		}
+		if rule.Identifier == "" {
+			return ast.RuleSet{}, fmt.Errorf("%s does not exist in the ruleset", ruleName)
+		}
+		ruleIdents := GetRuleIdentifiers(rule)
+		for ident := range ruleIdents {
+			// Get Imports
+			if sliceContains(ident, yaraModules) {
+				dependencies.Imports = append(dependencies.Imports, ident)
+			} else {
+				// Get Rules
+				for _, rule := range ruleset.Rules {
+					if rule.Identifier == ident {
+						dependencies.Rules = append(dependencies.Rules, rule)
+						break
+					}
+				}
+			}
+		}
+	}
+	return dependencies, nil
+}
+
+// GetRuleIdentifiers will find all the identifiers (excluding ForLoop variables)
+// and the number of times each identifier is seen for a given YARA rule
+func GetRuleIdentifiers(rule ast.Rule) map[string]int {
+	ruleIdentifiers := make(map[string]int)                   // ruleIdentifiers contains [identifier]numOfTimesSeen
+	queue := append([]queueT{}, queueT{node: rule.Condition}) // queue contains all the nodes to be processed
+	for len(queue) > 0 {
+		queueItem := &queue[0]
+		if _, ok := queueItem.node.(*ast.ForIn); ok {
+			// ForIn node found, extract loop variables and add them to the ignoreList
+			varsToIgnore := queueItem.node.(*ast.ForIn).Variables
+			queueItem.ignoreList = append(queueItem.ignoreList, varsToIgnore...)
+		}
+		addNodeIdentifierToIdentifiersMap(&queue, ruleIdentifiers)
+		addNodeChildrenToQue(&queue)
+		queue = append(queue[:0], queue[1:]...) // Delete node from queue
+	}
+	return ruleIdentifiers
+}
+
+// addNodeChildrenToQue adds a nodes children to the queue
+func addNodeChildrenToQue(queue *[]queueT) {
```

Comment:
By looking at this function's signature or documentation is not clear what it does. The name implies that the children of some node are added the queue, but which node? By looking at the implementation is clear that its the children of the first node in the queue, but it requires you to read the function's code. 

```
// Adds node's children to queue.
addChildrenToQueue(node ast.Node, queue *[]queueT)
```